### PR TITLE
Issue#16 main page: add pagination to main page, add functions to add/delete favorites movies to local storage

### PR DESCRIPTION
**Pull request by myroslav-kolomiiets**

### CODE REVIEW

File: src/components/MainPage.js [hunk 1]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
   componentDidMount() {
-    this.props.getMovies.movies();
+    this.props.getMovies.movies(null, 1);
+    this.setState({ activePage: 1 });
+    localStorage.setItem('favorites', 'init');
   }
 
   handleSearchChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
     this.setState({ inputValue: event.target.value });
+    this.setState({ activePage: 1 });
```
Comment (zonzujiro) [inline src/components/MainPage.js]: This can be combined in one `setState`

File: src/components/MainPage.js [hunk 2]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
-  ) => {
-    event.preventDefault();
-    console.log('handleBottomNavigationChange');
-  };
-
-  openListOfFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('openListOfFavorites');
+  handlePageChange = pageNumber => {
+    const value = this.state.inputValue;
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `const { inputValue } = this.state`

File: src/components/MainPage.js [hunk 3]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
-    event: SyntheticInputEvent<HTMLInputElement>
-  ) => {
-    event.preventDefault();
-    console.log('handleBottomNavigationChange');
-  };
-
-  openListOfFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('openListOfFavorites');
+  handlePageChange = pageNumber => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Add types pls

File: src/components/MainPage.js [hunk 4]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+    const value = this.state.inputValue;
+    if (value) {
+      this.setState({ activePage: pageNumber });
+      this.props.getMovies.movies(value, pageNumber);
+    } else {
+      this.setState({ activePage: pageNumber });
```
Comment (zonzujiro) [inline src/components/MainPage.js]: DRY

File: src/components/MainPage.js [hunk 5]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
-    event.preventDefault();
-    console.log('handleBottomNavigationChange');
-  };
-
-  openListOfFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('openListOfFavorites');
+  handlePageChange = pageNumber => {
+    const value = this.state.inputValue;
+    if (value) {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Can be replaced with ternary, something like

`const value = this.state.currentValue || null`

File: src/components/MainPage.js [hunk 6]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+    } else {
+      this.setState({ activePage: pageNumber });
+      this.props.getMovies.movies(null, pageNumber);
+    }
   };
 
-  addMovieToListOfFavorites = () => {
-    console.log('addMovieToListOfFavorites');
+  addMovieToListOfFavorites = id => {
+    let favorites = localStorage.getItem('favorites');
```
Comment (zonzujiro) [inline src/components/MainPage.js]: I think you should use `const` here

File: src/components/MainPage.js [hunk 7]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+      this.props.getMovies.movies(value, pageNumber);
+    } else {
+      this.setState({ activePage: pageNumber });
+      this.props.getMovies.movies(null, pageNumber);
+    }
   };
 
-  addMovieToListOfFavorites = () => {
-    console.log('addMovieToListOfFavorites');
+  addMovieToListOfFavorites = id => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Types

File: src/components/MainPage.js [hunk 8]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+    let favorites = localStorage.getItem('favorites');
+    if (favorites) {
+      let res = favorites.split(',');
```
Comment (zonzujiro) [inline src/components/MainPage.js]: You taking value from `localStorage`. Pls, replace with `JSON.parse`.

File: src/components/MainPage.js [hunk 9]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+      let duplicate = res.find(function(element) {
+        return element === id.toString();
+      });
+      if (!duplicate) {
+        res.push(id.toString());
+        let favToString = res.join(',');
+        localStorage.setItem('favorites', favToString);
+      }
+    }
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Ok, let's try functional composition here.

Pls, replace this part with helper with `JSON.parse`, `flow`, `find` from lodash/fp.

File: src/components/MainPage.js [hunk 10]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
   };
 
-  deleteMovieFromFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('deleteMovieFromFavorites');
+  deleteMovieFromFavorites = id => {
+    let favorites = localStorage.getItem('favorites');
+    if (favorites) {
+      let res = favorites.split(',');
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `JSON.parse`

File: src/components/MainPage.js [hunk 11]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+      let itemToRemove = res.indexOf(id.toString());
+      if (itemToRemove > -1) {
+        res.splice(itemToRemove, 1);
+        let favToString = res.join(',');
```
Comment (zonzujiro) [inline src/components/MainPage.js]: All of this can be replaced with `_.remove`

File: src/components/MainPage.js [hunk 12]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
 
-  deleteMovieFromFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('deleteMovieFromFavorites');
+  deleteMovieFromFavorites = id => {
+    let favorites = localStorage.getItem('favorites');
+    if (favorites) {
+      let res = favorites.split(',');
+      let itemToRemove = res.indexOf(id.toString());
+      if (itemToRemove > -1) {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Always use newer methods. JS have `Array#find` for such stuff

File: src/components/MainPage.js [hunk 13]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+        let favToString = res.join(',');
+        localStorage.setItem('favorites', favToString);
+      }
+    }
   };
 
   render() {
     const { classes, data, error } = this.props;
     const poster = 'https://image.tmdb.org/t/p/w500';
+    const { total_pages } = data.data;
```
Comment (zonzujiro) [inline src/components/MainPage.js]: You know, I don't like naming `data.data`. Can you change it?

File: src/components/MainPage.js [hunk 14]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+        let favToString = res.join(',');
+        localStorage.setItem('favorites', favToString);
+      }
+    }
   };
 
-  deleteMovieFromFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('deleteMovieFromFavorites');
+  deleteMovieFromFavorites = id => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Types, if they needed

File: src/components/MainPage.js [hunk 15]
```diff
@@ -59,48 +56,70 @@ type Props = {
...
+  activePage: number,
+  favorites: string,
 };
 
 class MainPage extends Component<Props, State> {
   componentDidMount() {
-    this.props.getMovies.movies();
+    this.props.getMovies.movies(null, 1);
+    this.setState({ activePage: 1 });
+    localStorage.setItem('favorites', 'init');
```
Comment (zonzujiro) [inline src/components/MainPage.js]: I think pagination must be handled thru url, some kind of `?page=1`, etc.

File: src/modules/movies.js [hunk 16]
```diff
@@ -28,33 +28,36 @@ export default function reducer(state: Object = initialState, action: Object) {
...
   return { type: MOVIE_LOADED, data: json, isLoaded: true };
 }
 
 export function handleErrors(error: Object) {
   return { type: MOVIE_ERROR, data: error, isLoaded: false };
 }
 
-export function movies(value: string | number) {
+export function movies(value: string | number, page: number) {
   if (value) {
```
Comment (zonzujiro) [inline src/modules/movies.js]: For such cases, better make two functions and then call them separately. 

In what cases we can call this function and not pass `value` to it?

**Review by zonzujiro — CHANGES_REQUESTED** [review]
LGTM, but requires some minor changes.

Also, maybe better will be replacing parts with uniq elements of an array with `Set`? 🤔

File: src/components/MainPage.js (hunk 1 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 2 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 3 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 4 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Fixed

File: src/components/MainPage.js (hunk 5 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 6 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 7 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 8 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 9 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 10 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 12 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 13 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Changed

File: src/components/MainPage.js (hunk 14 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Added

File: src/components/MainPage.js (hunk 15 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/modules/movies.js (hunk 16 above)
Comment (myroslav-kolomiiets) [reply src/modules/movies.js]: We call this function without `value` when we load main page on first page load, as default.

File: src/components/MainPage.js [hunk 17]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
   inputValue: number | string,
-  favorites: Array<number>,
+  activePage: number,
+  favorites: string,
 };
 
 class MainPage extends Component<Props, State> {
   componentDidMount() {
-    this.props.getMovies.movies();
+    const urlParamsValues = queryString.parse(this.props.location.search);
```
Comment (zonzujiro) [inline src/components/MainPage.js]: AFAIK `react-router` have such functionality which is called `match`.

If no, you can get this data with `URLSearchParams `?

File: src/components/MainPage.js [hunk 18]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
+    this.setState({ activePage: urlParamsValues.page || 1 });
+    this.props.getMovies.movies(
+      null,
+      urlParamsValues.page || 1,
+      'popularity.desc'
+    );
+    const favorites = localStorage.getItem('favorites') || JSON.stringify([]);
+    localStorage.setItem('favorites', favorites);
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `localStorage.setItem('favorites', localStorage.getItem('favorites'))` right? 🤔

File: src/components/MainPage.js [hunk 19]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
 
 class MainPage extends Component<Props, State> {
   componentDidMount() {
-    this.props.getMovies.movies();
+    const urlParamsValues = queryString.parse(this.props.location.search);
+    this.setState({ activePage: urlParamsValues.page || 1 });
+    this.props.getMovies.movies(
+      null,
+      urlParamsValues.page || 1,
+      'popularity.desc'
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Can you put this string in to constant with a descriptive name?

File: src/components/MainPage.js [hunk 20]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
 
-  //TODO: add functionality to handle list of favorites movies and bottom navigation
-  handleBottomNavigationChange = (
-    event: SyntheticInputEvent<HTMLInputElement>
-  ) => {
-    event.preventDefault();
-    console.log('handleBottomNavigationChange');
+  handlePageChange = (pageNumber: number) => {
+    const { inputValue } = this.state;
+    const value = inputValue || null;
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `const { inputValue = null } = this.state;` should work

File: src/components/MainPage.js [hunk 21]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
+    this.setState({ activePage: pageNumber });
+    this.props.getMovies.movies(value, pageNumber);
+    const history = createBrowserHistory({ basename: '/' });
```
Comment (zonzujiro) [inline src/components/MainPage.js]: You already have `history` API in a browser and in RR. Why you need third-party package?

File: src/components/MainPage.js [hunk 22]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
 
-  openListOfFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('openListOfFavorites');
+  addMovieToListOfFavorites = (id: number) => {
+    const favorites = localStorage.getItem('favorites');
+    if (favorites) {
+      const parsedFavorites = JSON.parse(favorites);
+      const duplicate = _.find(e => e === id, parsedFavorites);
+      if (duplicate !== id) {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: I think you need `contains` here

File: src/components/MainPage.js [hunk 23]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
+        parsedFavorites.push(id);
+        const favToString = JSON.stringify(parsedFavorites);
+        localStorage.setItem('favorites', favToString);
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `localStorage.setItem('favorites', JSON.stringify([...parsedFavorites, id]))`

File: src/components/MainPage.js [hunk 24]
```diff
@@ -51,60 +52,99 @@ const styles = () => ({
...
 
-  addMovieToListOfFavorites = () => {
-    console.log('addMovieToListOfFavorites');
+  deleteMovieFromFavorites = (id: number) => {
+    let favorites = localStorage.getItem('favorites');
+    if (favorites) {
+      const parsedFavorites = JSON.parse(favorites);
+      const updatedList = _.filter(e => e !== id, parsedFavorites);
+      const updatedListToString = JSON.stringify(updatedList);
+      localStorage.setItem('favorites', updatedListToString);
```
Comment (zonzujiro) [inline src/components/MainPage.js:130]: Try to use `_.flow` here - composition. You will avoid creation of redundant variables

File: src/modules/movies.js [hunk 25]
```diff
@@ -28,33 +28,38 @@ export default function reducer(state: Object = initialState, action: Object) {
...
+export function receiveMovies(json: Object) {
   return { type: MOVIE_LOADED, data: json, isLoaded: true };
 }
 
 export function handleErrors(error: Object) {
   return { type: MOVIE_ERROR, data: error, isLoaded: false };
 }
 
-export function movies(value: string | number) {
+export function movies(value: string | number, page: number, sort: string) {
```
Comment (zonzujiro) [inline src/modules/movies.js]: `sortBy` maybe? 🤔

File: src/modules/movies.js [hunk 26]
```diff
@@ -28,33 +28,38 @@ export default function reducer(state: Object = initialState, action: Object) {
...
         dispatch(handleErrors(error));
       }
     };
   } else {
     return async (dispatch: Dispatch) => {
       try {
-        const movies = await api.discoverMovie();
-        console.log(movies);
-        return dispatch(receiveMovies(movies.results));
+        const searchParams = {
```
Comment (zonzujiro) [inline src/modules/movies.js]: DRY for `searchParams`

File: src/modules/movies.js [hunk 27]
```diff
@@ -28,33 +28,38 @@ export default function reducer(state: Object = initialState, action: Object) {
   }
 }
 
-export function receiveMovies(json: Array<number | string> = []) {
+export function receiveMovies(json: Object) {
```
Comment (zonzujiro) [inline src/modules/movies.js]: Pls, replace it with arrow functions. ES6, etc.

File: src/components/MainPage.js [hunk 28]
```diff
@@ -176,31 +240,29 @@ class MainPage extends Component<Props, State> {
...
     }
   }
 }
 
 function mapStateToProps(state) {
-  return { data: state.movies, isLoaded: state.isLoaded };
+  return { fetchedDataBundle: state.movies, isLoaded: state.isLoaded };
 }
 
 const mapDispatchToProps = (dispatch: Dispatch<>) => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: I just saw - `Dispatch<>` is a very weird type :) I think must be `Dispatch<Action>`.

Or add flow-types for `react-redux` and a `redux`

File: src/components/MainPage.js [hunk 29]
```diff
@@ -152,22 +204,34 @@ class MainPage extends Component<Props, State> {
...
+                        </Typography>
                       </CardContent>
                       <CardActions>
                         <IconButton
                           className={classes.button}
                           aria-label="Delete"
-                          disabled
                           color="primary"
-                          onClick={this.deleteMovieFromFavorites}
+                          onClick={this.deleteMovieFromFavorites.bind(
```
Comment (zonzujiro) [inline src/components/MainPage.js]: No. This is bad. You will recreate this function on every render. Also, you will update props of this component on every render.

Pls, use a class method instead here and in all other places in render.

**Review by zonzujiro — CHANGES_REQUESTED** [review]
And one more time 👍

File: src/components/MainPage.js (hunk 18 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: I was rewrite it in more descriptive way

File: src/components/MainPage.js (hunk 19 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 20 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 21 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: It works fine in browser, but I can't successfully pass test's in Travis, using browser history API, that's why I add third-party package to project

File: src/components/MainPage.js (hunk 23 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/modules/movies.js (hunk 25 above)
Comment (myroslav-kolomiiets) [reply src/modules/movies.js]: Done

File: src/modules/movies.js (hunk 26 above)
Comment (myroslav-kolomiiets) [reply src/modules/movies.js]: Done

File: src/modules/movies.js (hunk 27 above)
Comment (myroslav-kolomiiets) [reply src/modules/movies.js]: Done

File: src/components/MainPage.js (hunk 28 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js (hunk 29 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Fixed

File: src/components/MainPage.js (hunk 22 above)
Comment (myroslav-kolomiiets) [reply src/components/MainPage.js]: Done

File: src/components/MainPage.js [hunk 30]
```diff
@@ -51,60 +52,102 @@ const styles = () => ({
...
+    history.push(`?page=${pageNumber}`);
   };
 
-  openListOfFavorites = (event: SyntheticInputEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    console.log('openListOfFavorites');
+  addMovieToListOfFavorites = (id: number) => () => {
+    const favorites = localStorage.getItem('favorites');
+    if (favorites) {
+      const hasDuplicate = _.flow([_.contains(id)])(favorites);
```
Comment (zonzujiro) [inline src/components/MainPage.js:114]: Flow for one function? Realy? 😄 I think just `contains` will work.

Also, you can remove this var and just right:

`if (_.contains(id, favourites))`

File: src/components/MainPage.js (hunk 24 above)
Comment (zonzujiro) [reply src/components/MainPage.js:130]: 👀

File: src/modules/movies.js [hunk 31]
```diff
@@ -26,38 +26,45 @@ export default function reducer(state: Object = initialState, action: Object) {
...
+};
 
-export function receiveMovies(json: Array<number | string> = []) {
+export const receiveMovies = (json: Object) => {
   return { type: MOVIE_LOADED, data: json, isLoaded: true };
-}
+};
 
-export function handleErrors(error: Object) {
+export const handleErrors = (error: Object) => {
```
Comment (zonzujiro) [inline src/modules/movies.js:35]: An error have type `Error`

File: src/modules/movies.js [hunk 32]
```diff
@@ -13,7 +13,7 @@ const MOVIE_ERROR = 'MOVIE_ERROR';
 
 //TODO: add some concrete types for the state and for actions.
 
-export default function reducer(state: Object = initialState, action: Object) {
+const reducer = (state: Object = initialState, action: Object) => {
```
Comment (zonzujiro) [inline src/modules/movies.js:16]: Our state and action have types. Can you use it here?

File: src/components/MainPage.js [hunk 33]
```diff
@@ -152,22 +207,28 @@ class MainPage extends Component<Props, State> {
...
+                        </Typography>
                       </CardContent>
                       <CardActions>
                         <IconButton
                           className={classes.button}
                           aria-label="Delete"
-                          disabled
                           color="primary"
-                          onClick={this.deleteMovieFromFavorites}
+                          onClick={this.deleteMovieFromFavorites(item.id)}
```
Comment (zonzujiro) [inline src/components/MainPage.js:223]: You creating a new function on every render. Pls, don't 😄 

You can handle such situations in two ways. Frist one - you can make a new component from this `Grid` and pass inside of it all props from `item` and handlers.
Or you can add some `data-*` attribute to this `IconButton` and try to get id from this attribute.

First one - is more "React'ish" and a second - easier to implement.

**Review by zonzujiro — CHANGES_REQUESTED** [review]
Use the force, Luke ⭐️

**Pull request closed without merging**